1/n Move precision plugin into strategy - update reference

four4fish opened the pull request:
## What does this PR do?

**1/n Move Precision Plugin into Strategy and update references**
2/n Move optimizer, lr to Strategy, remove accelerator routing functions 
3/n Simplify and reduce hooks and redundancies

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Fixes #7324 which is part of #10416

### Does your PR introduce any breaking changes? If yes, please list them.

<!-- FILL IN or None -->

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

kaushikb11 commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +44,18 @@ class Accelerator:
     """
 
-    def __init__(self, precision_plugin: PrecisionPlugin, training_type_plugin: TrainingTypePlugin) -> None:
+    def __init__(self, training_type_plugin: TrainingTypePlugin, precision_plugin: PrecisionPlugin = None) -> None:
```
```suggestion
    def __init__(self, training_type_plugin: TrainingTypePlugin, precision_plugin: Optional[PrecisionPlugin] = None) -> None:
```

ananthsub commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +44,20 @@ class Accelerator:
+
         self.training_type_plugin = training_type_plugin
 
+        if precision_plugin:
+            self.training_type_plugin._precision_plugin = precision_plugin
```
should this include a deprecation message for passing precision plugin this way?

ananthsub reviewed (commented):
It'd be very good to list what APIs this lets us simplify or remove from the Accelerator/strategy/precision plugin. Those could be listed as immediate followups to this PR

awaelchli commented on pytorch_lightning/plugins/training_type/training_type_plugin.py:
```diff
@@ -33,16 +34,27 @@ class TrainingTypePlugin(ABC):
+        return self._precision_plugin.precision
+
+    @property
+    def precision_plugin(self) -> Optional[PrecisionPlugin]:
```
```suggestion
    def precision_plugin(self) -> PrecisionPlugin:
```
I believe this should not be optional given the instantiation above.

awaelchli replied:
fixes mypy

awaelchli commented on pytorch_lightning/trainer/connectors/accelerator_connector.py:
```diff
@@ -407,6 +406,7 @@ def training_type_plugin(self) -> TrainingTypePlugin:
         # attach checkpoint plugin to the training type plugin
         if self._checkpoint_io is not None:
             self._training_type_plugin.checkpoint_io = self._checkpoint_io
+        self._training_type_plugin._precision_plugin = self.precision_plugin
```
Perhaps we need to None check here too?

four4fish replied:
If it's none, in training_type_plugin init() it will be set to Precision_plugin(), should we raise misconfig() or runtimeError if precision_plugin is None though? I feel just default Precision_plugin() could cause silent failures?

awaelchli replied:
Actually it can never be None. In the AcceleratorConnector.select_precision_plugin we always return a plugin and have a misconfiguration when no precision is detected. So we should be fine here.

awaelchli commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -1669,7 +1669,7 @@ def amp_backend(self) -> Optional[str]:
     @property
     def precision(self) -> Union[str, int]:
-        return self.accelerator.precision
+        return self.training_type_plugin.precision
```
```suggestion
        return self.training_type_plugin.precision_plugin.precision
```
in case we want to skip the property definition as shared in my comment above.

awaelchli commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +44,20 @@ class Accelerator:
+    ) -> None:
         """
         Args:
             precision_plugin: the plugin to handle precision-specific parts
```
note to self: update

awaelchli commented on pytorch_lightning/plugins/training_type/sharded_spawn.py:
```diff
@@ -118,7 +118,7 @@ def post_training_step(self):
         # Ensure that the scaler points to the correct process group
         # which is re-initialized in a new process
-        precision_plugin = trainer.accelerator.precision_plugin
+        precision_plugin = trainer.precision_plugin
         if isinstance(precision_plugin, ShardedNativeMixedPrecisionPlugin):
             precision_plugin.scaler = ShardedGradScaler()
```
```suggestion
        if isinstance(self.precision_plugin, ShardedNativeMixedPrecisionPlugin):
            self.precision_plugin.scaler = ShardedGradScaler()
```

awaelchli commented on pytorch_lightning/plugins/training_type/tpu_spawn.py:
```diff
@@ -56,11 +57,14 @@ def __init__(
         self,
         parallel_devices: Optional[List[int]] = None,
         checkpoint_io: Optional[CheckpointIO] = None,
+        precision_plugin: Optional[PrecisionPlugin] = None,
```
note to self: update trainer.precision_plugin references here

four4fish commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -248,26 +253,29 @@ def setup_training_type_plugin(self) -> None:
     @property
     def precision(self) -> Union[str, int]:
-        return self.precision_plugin.precision
+        """deprecated."""
```
Sorry for the poor wording, will update

four4fish commented on pytorch_lightning/plugins/training_type/training_type_plugin.py:
```diff
@@ -33,16 +34,27 @@ class TrainingTypePlugin(ABC):
     def checkpoint_io(self) -> CheckpointIO:
         return self._checkpoint_io
 
+    @property
+    def precision(self) -> Union[str, int]:
+        return self._precision_plugin.precision
```
I don't see too much value. trainer already have it as well, I think we can remove this?

awaelchli replied:
I think so too :)

kaushikb11 commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +45,29 @@ class Accelerator:
+            precision_plugin: the plugin to handle precision-specific parts
+
+                .. deprecated:: v1.6
+                    The ``precision_plugin`` parameter has been deprecated in v1.6 and will be removed in 1.7.
```
```suggestion
                    The ``precision_plugin`` parameter has been deprecated in v1.6 and will be removed in 1.8.
```
Should we remove it for v1.8 release instead?

awaelchli replied:
I'm not sure. A few thoughts:

- I'm not even sure we can uphold this deprecation/backward compatibility if we go all the way with https://docs.google.com/document/d/1E5t8auWf5DrNHzutvMmrJC_KqBqVyZuYX0thra69Ad8/edit#heading=h.nehmb825f40s
- We want accelerator/plugins to be stable at 1.6 - so we might as well introduce larger breaking changes now 
- So far we have always chosen a shorter deprecation of only one release for all plugin/accelerator related items.

RFC @ananthsub @four4fish

four4fish replied:
Yeah I agree, in step 3 move accelerator into TTP, there is no way we can keep backwards compatibility. Precision plugin has to be removed then. 

> We want accelerator/plugins to be stable at 1.6 - so we might as well introduce larger breaking changes now

Agree! how about "this is deprecated and will be removed soon"?

ananthsub replied:
The precision plugin would be removed from here once we invert the Strategy/Accelerator hierarchy. @four4fish I don't think the deprecation warning would last through that refactor.

If that's the case, then we don't need the warning here. We can add the basic support in the interim for people on master, but the Accelerator shouldn't accept a precision plugin when 1.6 is released, since the accelerator won't have a reference to the strategy/training type plugin by then.

four4fish replied:
Agree, will update.
I think as soon as the step 3: moving accelerator start, accelerator won't accept precision plugin anymore.

kaushikb11 commented on pytorch_lightning/plugins/training_type/training_type_plugin.py:
```diff
@@ -33,16 +34,23 @@ class TrainingTypePlugin(ABC):
         return self._checkpoint_io
 
+    @property
+    def precision_plugin(self) -> PrecisionPlugin:
```
Should we have a setter method for it as well?

four4fish replied:
`training_type_plugin._precision_plugin=`
There are overrides for precision_plugin(self) logic in deepspeed.

kaushikb11 commented:
Also, we would be breaking it by changing the signature of the `Accelerator` from

```python
def __init__(self, precision_plugin: PrecisionPlugin, training_type_plugin: TrainingTypePlugin)

# to

def __init__(self, training_type_plugin: TrainingTypePlugin, precision_plugin: Optional[PrecisionPlugin] = None)
```

four4fish commented:
> Also, we would be breaking it by changing the signature of the `Accelerator` from
> 
> ```python
> def __init__(self, precision_plugin: PrecisionPlugin, training_type_plugin: TrainingTypePlugin)
> 
> # to
> 
> def __init__(self, training_type_plugin: TrainingTypePlugin, precision_plugin: Optional[PrecisionPlugin] = None)
> ```

Yes it will for positional parameter passing in. But precision_plugin is optional option now, has to switch the order.  Soon we will do step 3 which is move Accelerator into strategy, we have to remove precision_plugin entirely to avoid circle import. So it will be breaking change

kaushikb11 commented on pytorch_lightning/trainer/connectors/accelerator_connector.py:
```diff
@@ -407,7 +407,9 @@ def training_type_plugin(self) -> TrainingTypePlugin:
             self._training_type_plugin.checkpoint_io = self._checkpoint_io
-        self._training_type_plugin._precision_plugin = self.precision_plugin
+        precision_plugin = self.precision_plugin
+        if precision_plugin is not None:
```
But it could never be None?

four4fish replied:
Yeah, as we discussed above. But consider all the refactor happening now, feel safer have a none check? 
Training_type_plugin(precision_plugin=None) will have default value, but training_type_plugin.precision_plugin will have not default value.

carmocca commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +45,29 @@ class Accelerator:
         self.training_type_plugin = training_type_plugin
 
+        if precision_plugin:
+            """
+            .. deprecated
+                precision_plugin parameter is deprecated will be removed soon.
+                Use :`training_type_plugin(precision_plugin) instead.
+            """
```
nit: Triple quotes should be for docstrings, which this is not. Can you use comments instead?

kaushikb11 replied:
Comments for deprecation notes?

awaelchli replied:
This must have been a merge resolution issue because I had already fixed this, including the formatting @four4fish.

awaelchli replied:
Yep, that's mostly it, but you can keep the text ".. deprecated soon" instead of v1.7 as the target :)

four4fish replied:
is 
```
 .. deprecated:: soon
                    The ``precision_plugin`` parameter has been deprecated and will be removed in soon
                    Pass the precision plugin as a parameter to the ``TrainingTypePlugin`` instead.
```
A valid format? Or do you think use comments better?

awaelchli replied:
Nope :) 

This is our standard format: 

```
.. deprecated:: v1.6
                    The ``precision_plugin`` parameter has been deprecated in v1.6 and will be removed soon
                    Pass the precision plugin as a parameter to the ``TrainingTypePlugin`` instead.
```

four4fish replied:
How about we just use comments then? we gonna remove before 1.6, don't want to misleading user

awaelchli replied:
Comments don't show in the documentation. We dont expect them to read the source files, right?
We could just drop the version number

.. deprecated::
    The ``precision_plugin`` parameter has been deprecated and will be removed soon
    Pass the precision plugin as a parameter to the ``TrainingTypePlugin`` instead.

carmocca commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +45,29 @@ class Accelerator:
 
-    def __init__(self, precision_plugin: PrecisionPlugin, training_type_plugin: TrainingTypePlugin) -> None:
+    def __init__(
+        self, training_type_plugin: TrainingTypePlugin, precision_plugin: Optional[PrecisionPlugin] = None
```
This change is not backward compatible, as the order is reversed now compared to before.

The only solution would be to set a token default value for training type plugin.

On a tangential note, I don't get the point for deprecating just the constructor, either the API is stable or it isn't, doing it for the constructor seems arbitrary.

awaelchli replied:
Reviewers need to agree here or we can't move forward. 

Either we add deprecation and backward compatibility for a few commits for all the 0.0001% users who actually install master and use these APIs, OR we remove the reference here, breaking change, and move on.

> either the API is stable or it isn't

The API is not stable and we know that from the docs. We prioritize making it stable for 1.6.

tchaton replied:
Hey @awaelchli. Adding to @carmocca  point, let's add the deprecation warning and not change the init API. Better to do it at once when we actually break the API.

carmocca replied:
Sorry if this wasn't clear, I'm fine with both deprecating OR not deprecating, I was just raising a question. I don't want to block the PR over that.

but IF the deprecation is chosen, then my point about setting a token default value stands.

awaelchli replied:
@carmocca Is the current way sufficient?

carmocca commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -248,26 +263,36 @@ def setup_training_type_plugin(self) -> None:
+
+        Use `training_type_plugin.precision_plugin.precision` instead.
+        """
+        rank_zero_deprecation(
+            f"`{self.__class__.__name__}.precision` was and will be removed soon"
+            f" Use `training_type_plugin.precision_plugin.precision` instead."
+        )
```
Can you merge/rebase master into this PR? To avoid any potential failures after #9940

four4fish replied:
Checked

carmocca commented on pytorch_lightning/plugins/training_type/deepspeed.py:
```diff
@@ -456,8 +459,7 @@ def init_deepspeed(self):
                 "DeepSpeed currently does not support different `accumulate_grad_batches` at different epochs."
             )
 
-        precision = self.lightning_module.trainer.accelerator.precision
```
awesome!

carmocca commented on pytorch_lightning/trainer/trainer.py:
```diff
@@ -1568,7 +1568,7 @@ def training_type_plugin(self) -> TrainingTypePlugin:
 
     @property
     def precision_plugin(self) -> PrecisionPlugin:
-        return self.accelerator.precision_plugin
+        return self.training_type_plugin.precision_plugin
```
Related to my comment about it being Optional, if that's the case, then this should check `None` and the typing would need to be fixed. Same for `precision` below

four4fish replied:
It won't be None as 1. in accelerator we make sure precision plugin is not None (there was a resolved conversation about about this) 2. in ttp initialization, if precision_plugin is none, it will auto assign precision_plugin()

carmocca commented on pytorch_lightning/plugins/training_type/ddp.py:
```diff
@@ -86,6 +87,7 @@ def __init__(
         parallel_devices: Optional[List[torch.device]] = None,
         cluster_environment: Optional[ClusterEnvironment] = None,
         checkpoint_io: Optional[CheckpointIO] = None,
+        precision_plugin: Optional[PrecisionPlugin] = None,
```
Why is it `Optional` here whereas it was required for the `Accelerator`?

This comment also applies to all other additions to each TTP.

four4fish replied:
Optional for user who construct Strategy(), they can not pass percision_plugin which default is None. Then we set to default for them. If it's TPU or IPU we know which precision_plugin to set when it's None, if not we use precision_plugin().

If the precision plugin choiced by accelerator_connector, it guaranteed won't be None

tchaton commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -44,15 +45,29 @@ class Accelerator:
         self.training_type_plugin = training_type_plugin
 
+        if precision_plugin:
+            """
```
Let's move the depreciation within the docstring of the init function.

tchaton commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -159,31 +174,31 @@ def training_step(self, step_kwargs: Dict[str, Union[Any, int]]) -> STEP_OUTPUT:
         See :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step` for more details
         """
-        with self.precision_plugin.train_step_context():
+        with self.training_type_plugin.precision_plugin.train_step_context():
```
Should the precision plugin own the `step_context` instead? I don't have a strong opinion there.

four4fish replied:
Yep, all steps will be either move to training_type_plugin or precision in 2/n or 3/n

tchaton commented on pytorch_lightning/accelerators/accelerator.py:
```diff
@@ -84,7 +99,7 @@ def pre_dispatch(self, trainer: "pl.Trainer") -> None:
             self.setup_optimizers(trainer)
 
-        self.precision_plugin.pre_dispatch()
+        self.training_type_plugin.precision_plugin.pre_dispatch()
```
Should we use strategy already instead ?

four4fish replied:
Agree we should unify the term. we haven't make that change yet? How about address in a different PR?

awaelchli replied:
Yes, the renaming has to happen in a separate PR and it will touch many files. This can definitely not be done here.

tchaton commented on pytorch_lightning/trainer/connectors/accelerator_connector.py:
```diff
@@ -407,6 +407,9 @@ def training_type_plugin(self) -> TrainingTypePlugin:
         # attach checkpoint plugin to the training type plugin
         if self._checkpoint_io is not None:
             self._training_type_plugin.checkpoint_io = self._checkpoint_io
+        precision_plugin = self.precision_plugin
+        if precision_plugin is not None:
+            self._training_type_plugin._precision_plugin = self.precision_plugin
```
```suggestion
        if self.precision_plugin is not None:
            self._training_type_plugin._precision_plugin = self.precision_plugin
```

four4fish replied:
There is selection logic build in property precision plugin(): It's better call it once instead multiple times

https://github.com/PyTorchLightning/pytorch-lightning/blob/c5cabf836e0729bd48c16f538d48bb72651a73b8/pytorch_lightning/trainer/connectors/accelerator_connector.py#L393-L397

The self.select_precision_plugin() is quite heavy logic

awaelchli replied:
Sounds good @four4fish , just a note, it is currently also used multiple times, so you could make it also 

```suggestion
        precision_plugin = self.precision_plugin
        if precision_plugin is not None:
            self._training_type_plugin._precision_plugin = precision_plugin
```

four4fish replied:
Yeah! typo!! thank you for catching it :)

tchaton commented on pytorch_lightning/trainer/connectors/accelerator_connector.py:
```diff
@@ -795,12 +798,10 @@ def select_accelerator(self) -> Accelerator:
+        accelerator = acc_cls(training_type_plugin=self.training_type_plugin)
         # transfer ownership of the plugins to the accelerator
         self._training_type_plugin = proxy(self.training_type_plugin)
-        self._precision_plugin = proxy(self.precision_plugin)
```
Neat !

four4fish commented:
@tchaton @carmocca Addressed the comments, could you guys take a look?
1. Now this 1/n is backward compatible. 
2. Updated the deprecation message and format. Add to @awaelchli mentioned before: there will breaking change in refactor step3 "move accelerator into strategy". Which means we will have to remove precision_plugin param from Accelerator BEFORE we have a stable version to release. So the deprecation message is "will be deprecated soon" instead of "deprecated in 1.6 remove in 1.7"
